really simple dialog for selecting qss theme

glennlive opened the pull request:
Added a simple selection of theme from the grc menu... Talking with Tom it may be a useful feature to have, although probably not in the right menu / naming scheme.   The idea is to be able to quickly switch themes without needed to edit the conf file.

sdh11 commented:
I think this looks fine for the dialog. The next step should be to save the selected QSS theme using GRC's preferences.

sdh11 commented:
I'm not sure about this yet, but the preferences may be saved in two different locations. If they are, then we would need to figure out a way to consolidate them so that gr-qtgui will load the correct theme.

trondeau commented:
I've been thinking this might work well as an option on a per-file basis. We can also have it trigger the env variable, GR_CONF_QTGUI_QSS.

skoslowski commented:
Before I do a detailed review, some observations: This code does set gr.prefs() already. Since this is sort outside of GRC I'd put this under tools. No keyboard shortcut. Remove the print statements.

Maybe it would be nicer to do this on a per-flowgraph basis. Add a param that's only shown for QTGUI and have it override the default from gr.prefs() via the generated code. Right now, I think a simple text entry is sufficient. Anything more fancy (dropdown) is probably not worth the hassle.

(of course we can have both, too)

skoslowski commented:
Haha, looks like we're +2 on per-file already =)

sdh11 commented:
I was also wondering if it would fit better under Tools. 

Doing a per-file version would require a change in qtgui as well, not just GRC. It's do-able, just not as easy as setting the config under gr.prefs().

sdh11 commented:
Nevermind. Looks like we can set it by calling tb.setStyleSheet() in python.

glennlive commented:
These all sound good.  Let me know if you want me to do anything.

On Aug 31, 2015 3:04 PM, sdh11 <notifications@github.com> wrote:
Nevermind. Looks like we can set it by calling tb.setStyleSheet() in python.

---
Reply to this email directly or view it on GitHub:
https://github.com/gnuradio/gnuradio-wg-grc/pull/17#issuecomment-136468517

skoslowski commented:
Sure, you can go ahead and implement a per-file setting by adding a param to options.xml and read the value in python/Generator.py, pass it to the template and have it generate the setter (if the value is not empty). Also, you can make the changes I mentioned above to your existing code.

(Oh, and rebase your branch to current master_grcwg please)

glennlive commented:
So from the above feedback, these features should be implemented now:

1. Stylesheets per flowgraph via option-block param.  
 * Drop down box with standard options for dark / alt / projector themes
  * The drop down also lets you enter a path / filename manually
  * Added a "Light" theme, but haven't gotten my stylesheet looking nice yet, so maybe should be removed from the list
 * Parameter passes through to Cheetah's template for the top-block (although I didn't add an explicit setter here, as it seemed like it wasn't needed, thoughts?)
 * Selecting "default" leaves the option blank, and the gr.prefs stylesheet is used

2. Moved dialog to Tools menu
 * Perhaps this can even go away if a full-blown options dialog is added for gr.prefs
 * Naming of menu still tentative ("Select QT Theme...")

3. Git Rebased.
 * I think.  That was more of a learning curve then expected.
 * Please let me know if I merged things ok

skoslowski commented:
...sorry for dropping the ball here...

I rebased your branch and cleaned the code a little. See https://github.com/skoslowski/gnuradio-wg-grc/tree/grc-select-theme. You can use this this by using "git reset --hard ..." after you added my clone as a remote in git.

I was thinking maybe it would be better to use a param type like file_open instead of a drop-down. If we set the right default path users would immediately see the available themes so we won't have to hard-code them. What do you think?

For the param key i'd prefer qt_qss_theme (the override is implied). The param should be shown only in the qt_gui generate mode.

glennlive commented:
I've pulled your version and started in on the file_open instead of drop-down in the option block.  However, it seems like the <value> element doesn't process variables from the xml... how can we get gr.prefix() there to setup the default path?

Your param key name makes sense, dropped the "override".  

What about the tools / preferences menu selection? Do you have have plans to add a complete preference-configurations window (which would remove the need for a separate qt-qss-theme selection)?

glennlive commented:
Okay... the only thing I came up with is adding another parameter type to the options block of "file_open_qss" so that the default directory in that dialog box can be set to the themes path.  What am I missing?  It seems like overkill, but I didn't find any easy way to get the value into interpreted from the xml block into the default path.

Thoughts?

skoslowski commented:
I agree, a new param type is overkill. To set a default (in which the dialog opens) you can write
```
  <value>gr.prefix() + '/share/gnuradio/themes/dark.qss'</value>
```
in the param and add ```<import>from gnuradio import gr</import>``` at the top. Not pretty, but functional =)

I don't have plans for a "complete preference-configurations window" myself. If you wanted to work on this, I suggest to create a standalone QT app, which is called from GRC.

glennlive commented:
Ahhg!  Of course.  I was trying to get grc to expand gr.preifx in the text box.  I will give this a shot to make sure it opens the correct dir sill.

glennlive commented:
The down side here is the empty text box... 

1). If a user clears out the default gr.prefix+ then the default directory in the file_open isn't really all that useful anymore.
2). Since the default option itself isn't empty but rather a path to a directory... it looks like you have a theme selected, when really you don't (when there is just a path, no file).

Let me know if you want anything else, or if you just want to kill the branch.

trondeau commented:
Do you mean scrap the idea of just this implementation branch? I really like the idea of having this option easily presented to the user.

glennlive commented:
@Tom: Glad to hear you want it.  :)  

After looking at it again last night, I think I contaminated my master branch somewhere along the way (I recompiled my master and some of the theme select option was still there).  So, I need to clean it up for an easy merge for Sebastian.

Before cleaning that up, which do you all prefer?

1) . The "QT theme" text box in the options block to be filled in with "gr.prefix() + '/share/gnuradio/themes', with the down-sides listed above.

--  or --

2). The semi-overkill new param option of file_open_qss so that grc can hide the gr.prefix from the user.  This has the benefit of the correct default directory even for the empty text field.

I think this boils down to gui and code preferences.  So input welcome.  Maybe there is a third option I'm missing?

trondeau commented:
I was going to consider just assuming the location of the QSS themes and put that into the code directly, so not a user-selectable option. But that might be a problem for people making their own and not having permission to install them.

Maybe we add a qss_theme_dir option to the preferences files? Have it point to the prefix/share/gnuradio/themes but treat it as a comma-separated list of possible directories to find .qss files. Have GRC just use this preference option to pull in the locations. Note that this means a user can also set GR_CONF_QTGUI_QSS_THEME_DIR env variable to set this easily, too.

Might that work better?

skoslowski commented:
@glennlive , I don't want this dropped either.

Option 1) is nice, being withing the boundaries of current param XML capabilities. If we want a dynamic dropdown/combo, we'd have to fill in the options during load. Since this is the Options block after all, I think it could be justified to make an exception here. @glennlive, that would be in __init__() of grc/base/Param.py.
Multiple search paths would work here, too. In my option, this would have to be provided by some code in QT GUI. In GRC we'd only get a list of theme names (from a helper in QT GUI) and than use that in the generated to call some helper within the QT GUI to locate and set the actual file.

Option 2) Is indeed semi-overkill. Instead of a new type I'd rather extend the XML defs to add a <default_path> to read only by file_* type params.

It comes down to want we really want here: A file open or a combo box. In other words, do we want these themes to be installed in some path or do we want them to be some file we reference?

glennlive commented:
I've been fine with the file selection dialog (although I suppose a drop down would be okay) what has been really bugging me is the default directory showing up correctly in the gui.  I don't think the init of base/Param is the correct place, because that only give the value once... not the gui session replacement of "" with gr.prefix()+"/share/gnuradio/themes".  I finally got oriented enough to realize it was a simple matter of check the param.key in the gui/param file_open function.   Let me know what you think, at least from the gui side of things it seems clean enough.

I'm happy added other / custom directories for the themes default, but I'm not sure most users would find that terribly necessary.   I suppose the use case is they have out-of-tree themes?  

Finally, what do you mean by "helper in the QT GUI"?  I'm also afraid that my master branch is still contaminated, but hopefully that isn't a problem.  If not I will have to rebase again from your repo.

skoslowski commented:
I think the file selection dialog is good for now. For the implementation only suggested base/Params.py for the Dropdown Options. For the file dialog I was thinking to add an XML tag to specify the default path from the XML. However, I like your implementation. If at some point more file_open params need to set a default path it can be added then.

By helper in the QT Gui I meant basically two simple functions to get a list of theme names (for the GUI) and to set a theme based on its name (on run). But that's mute now anyway. We might still add theme override via a env variable (in flowgraph.tmpl). What do you think?

I have some minor comments that I will add line-by-line. Other than that this is about ready to merged I think. Great work! Thanks!

skoslowski commented on grc/python/flow_graph.tmpl:
```diff
@@ -356,6 +356,19 @@ if __name__ == '__main__':
     #end if
     tb.show()
 
+    #if $flow_graph.get_option('qt_qss_theme') is not None and len($flow_graph.get_option('qt_qss_theme')) > 0:
```
Wouldn't ```#if $flow_graph.get_option('qt_qss_theme')``` suffice?

glennlive replied:
Yep.  My brain was trapped in a pessimistically over-cautions things-aren't-working yet fog, and it was too late for coffee.

skoslowski commented on grc/python/flow_graph.tmpl:
```diff
@@ -356,6 +356,19 @@ if __name__ == '__main__':
+        qssf = "$flow_graph.get_option('qt_qss_theme')"
+        if os.path.split(qssf)[0] == "":
+           qssf = gr.prefix() + "/share/gnuradio/themes/" + qssf
+        if os.path.split(qssf)[-1] != "":
```
This is rather hard to read. How about using exists(), dirname() and join() from the os.path module make intent clearer.
Also, the path logic should be run during code generation (It might be even easier to move it into Generator.py).
 
And we could try to get a relative path (e.g. the style file is in the same folder or a sub-folder of the generated code)

glennlive replied:
Should be more readable now...  and I also found some cases where the logic didn't do the right thing when given invalid directories, so I added better checking.  

And assuming the ".qss" if it doesn't exist means that you can just type "projector" in for the theme and it does the right thing.  :)

Anything else?

skoslowski commented:
I added some more minor edits while review it. See
https://github.com/skoslowski/gnuradio-wg-grc/tree/grc-select-theme

Also moved the generated code around a bit to get a cleaner output.

If we auto-append '.qss' it should be during the code generation, not at run-time. Removed it for now.

All that's left is to squash the commits into two or three meaningful change-sets (please prefix the commit messages with 'grc:').  You mentioned you had problems with your repo; so I can do this for you, if that helps

glennlive commented:
If you would squash the commits that would be great.  The rest makes sense too.

On Sep 30, 2015 3:28 PM, Sebastian Koslowski <notifications@github.com> wrote:

I added some more minor edits while review it. See
https://github.com/skoslowski/gnuradio-wg-grc/tree/grc-select-theme

Also moved the generated code around a bit to get a cleaner output.

If we auto-append '.qss' it should be during the code generation, not at run-time. Removed it for now.

All that's left is to squash the commits into two or three meaningful change-sets (please prefix the commit messages with 'grc:'). You mentioned you had problems with your repo; so I can do this for you, if that helps

-
Reply to this email directly or view it on GitHub<https://github.com/gnuradio/gnuradio-wg-grc/pull/17#issuecomment-144514639>.

skoslowski commented:
Done and merged into master_grcwg. Will request a merge into mainline soon. Thanks again for your work on this.
